census/estimator/trainer/input.py

```diff
@@ -0,0 +1,142 @@
+
+Tutorial on wide and deep: https://www.tensorflow.org/tutorials/wide_and_deep/
+"""
+
+
+import tensorflow as tf
+
+import trainer.featurizer as featurizer
+import trainer.metadata as metadata
+
+
+def _decode_csv(line):
+    """Takes the string input tensor and returns a dict of rank-2 tensors."""
+
+    # Takes a rank-1 tensor and converts it into rank-2 tensor
+    # Example if the data is ['csv,line,1', 'csv,line,2', ..] to
+    # [['csv,line,1'], ['csv,line,2']] which after parsing will result in a
+    # tuple of tensors: [['csv'], ['csv']], [['line'], ['line']], [[1], [2]]
+    row_columns = tf.expand_dims(line, -1)
+    columns = tf.decode_csv(
+        row_columns, record_defaults=metadata.CSV_COLUMN_DEFAULTS)
+    features = dict(zip(metadata.CSV_COLUMNS, columns))
+
+    # Remove unused columns
+    unused_columns = set(metadata.CSV_COLUMNS) - \
+        {col.name for col in featurizer.INPUT_COLUMNS} - \
+        {metadata.LABEL_COLUMN}
+    for col in unused_columns:
+        features.pop(col)
+    return features
+
+
+def _parse_label_column(label_string_tensor):
+    """Parses a string tensor into the label tensor.
+
+    Args:
+      label_string_tensor: Tensor of dtype string. Result of parsing the CSV
+        column specified by LABEL_COLUMN.
+
+    Returns:
+      A Tensor of the same shape as label_string_tensor, should return
+      an int64 Tensor representing the label index for classification tasks,
+      and a float32 Tensor representing the value for a regression task.
+    """
+    # Build a Hash Table inside the graph
+    table = tf.contrib.lookup.index_table_from_tensor(
+        tf.constant(metadata.LABELS))
+
+    # Use the hash table to convert string labels to ints and one-hot encode
+    return table.lookup(label_string_tensor)
+
+
+def input_fn(filenames,
+             num_epochs=None,
+             shuffle=True,
+             skip_header_lines=0,
+             batch_size=200):
+    """Generates features and labels for training or evaluation.
+
+    This uses the input pipeline based approach using file name queue
+    to read data so that entire data is not loaded in memory.
+
+    Args:
+        filenames: [str] A List of CSV file(s) to read data from.
+        num_epochs: (int) how many times through to read the data. If None will
+          loop through data indefinitely
+        shuffle: (bool) whether or not to randomize the order of data. Controls
+          randomization of both file order and line order within files.
+        skip_header_lines: (int) set to non-zero in order to skip header lines in
+          CSV files.
+        batch_size: (int) First dimension size of the Tensors returned by input_fn
+
+    Returns:
+        A (features, indices) tuple where features is a dictionary of
+          Tensors, and indices is a single Tensor of label indices.
+    """
+    dataset = tf.data.TextLineDataset(filenames).skip(skip_header_lines).map(
+        _decode_csv)
```

Comment:
I would add the `num_parallel_calls` to map as well as the prefetch at the end.

Want to enforce the usage as as is should be default usage in my opinion.

Reply:
Done. Added it as a parameter also.